test/widgets/test_clock.py

```diff
@@ -0,0 +1,332 @@
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+# Widget specific tests
+
+import datetime
+import sys
+from importlib import reload
+from time import sleep
+
+import pytest
+
+import libqtile.config
+from libqtile.bar import Bar
+from libqtile.widget import clock
+
+
+def no_op(*args, **kwargs):
+    pass
+
+
+class FakeWindow:
+    class _NestedWindow:
+        wid = 10
+    window = _NestedWindow()
+
+
+class MockDatetime(datetime.datetime):
+
+    @classmethod
+    def now(cls, *args, **kwargs):
+        return cls(2021, 1, 1, 10, 20, 30)
+
+    def astimezone(self, tzone=None):
+        if tzone is None:
+            return self
+        else:
+            return self + datetime.timedelta(hours=tzone)
+
+
+def test_clock(fake_qtile, monkeypatch):
+    """ test output with update (check number of updates and color) """
+    monkeypatch.setattr("libqtile.widget.clock.datetime", MockDatetime)
+    clk1 = clock.Clock()
+    fakebar = Bar([clk1], 24)
+    fakebar.window = FakeWindow()
+    fakebar.width = 10
+    fakebar.height = 10
+    fakebar.draw = no_op
+    clk1._configure(fake_qtile, fakebar)
+    text = clk1.poll()
+    assert text == "10:20"
+
+
+def test_clock_invalid_timezone(fake_qtile, monkeypatch, caplog):
+    """ test output with update (check number of updates and color) """
+    class FakeDateutilTZ:
+        @classmethod
+        def tz(cls):
+            return cls
+
+        @classmethod
+        def gettz(cls, val):
+            return None
+
+    # Stop system importing these modules in case they exist on environment
+    monkeypatch.setitem(sys.modules, "pytz", None)
+    monkeypatch.setitem(sys.modules, "dateutil", None)
+    monkeypatch.setitem(sys.modules, "dateutil.tz", None)
+
+    # Reload module to force ImportErrors
+    reload(clock)
+
+    # pytz and dateutil must not be in the sys.modules dict...
+    monkeypatch.delitem(sys.modules, "pytz")
+    monkeypatch.delitem(sys.modules, "dateutil")
+
+    # Override datetime
+    monkeypatch.setattr("libqtile.widget.clock.datetime", MockDatetime)
+
+    # Set up references to pytz and dateutil so we know these aren't being used
+    # If they're called, the widget would try to run None(self.timezone) which
+    # would raise an exception
+    clock.pytz = None
+    clock.dateutil = FakeDateutilTZ
+
+    # Fake datetime module just adds the timezone value to the time
+    clk2 = clock.Clock(timezone="1")
+
+    fakebar = Bar([clk2], 24)
+    fakebar.window = FakeWindow()
+    fakebar.width = 10
+    fakebar.height = 10
+    fakebar.draw = no_op
+    clk2._configure(fake_qtile, fakebar)
+
+    # An invalid timezone current causes a TypeError
+    with pytest.raises(TypeError):
+        clk2.poll()
+
+    # Default time is 10:20 and we add 1 hour for the timezone
+    assert "can not infer" in caplog.text
```

Comment:
Is it not possible to use `caplog` to check log output? Works on my machine but fails here.

I'll remove this line and just rely on the `TypeError`. That'll give us full coverage but it feels a little inelegant as I think the widget should handle the invalid timezone more gracefully than it does.